ofrak_core/ofrak/core/esp/app.py

```diff
@@ -0,0 +1,959 @@
+                break
+            
+            # Read segment header
+            segment_size = int.from_bytes(current_data[offset+4:offset+8], 'little')
+            offset += 8  # Skip segment header
+            
+            if offset + segment_size > len(current_data):
+                break
+            
+            # Collect segment data
+            segment_data = current_data[offset:offset+segment_size]
+            segment_data_list.append(segment_data)
+            offset += segment_size
+        
+        # Find checksum location (16-byte aligned) - use same formula as unpacker
+        checksum_offset = ((offset + 16) // 16) * 16 - 1
+        
+        # Calculate new checksum
+        calculated_checksum = ESP_APP_CHECKSUM_MAGIC  # 0xEF
+        for segment_data in segment_data_list:
+            for byte in segment_data:
+                calculated_checksum ^= byte
+        calculated_checksum &= 0xFF
+        
+        # Update checksum in the binary
+        if checksum_offset < len(current_data):
+            current_data[checksum_offset] = calculated_checksum
+        
+        # Calculate and update hash if needed (ESP32 with hash_appended)
+        hash_offset = checksum_offset + 1
+        if (has_extended_header and 
+            hash_offset + 32 <= len(current_data) and
+            current_data[23] == 1):  # hash_appended flag
+            
+            # Calculate SHA256 of data up to hash
+            image_data = current_data[:hash_offset]
+            calculated_hash = hashlib.sha256(image_data).digest()
+            
+            # Update hash in the binary
+            current_data[hash_offset:hash_offset+32] = calculated_hash
+        
+        # Update the main resource with the new data
+        resource.queue_patch(Range.from_size(0, len(current_data)), bytes(current_data))
+
+    async def verify_with_esptool(self, resource: Resource) -> bool:
+        """
+        Verify the packed ESP app using esptool image_info command.
+        
+        :param resource: The packed ESP app resource
+        :return: True if verification passes, False otherwise
+        """
+        try:
+            # Write the packed data to a temporary file
+            packed_data = await resource.get_data()
+            
+            with NamedTemporaryFile(suffix=".bin", delete=False) as temp_file:
+                temp_file.write(packed_data)
+                temp_file.flush()
+                
+                # Use esptool to verify the image
+                from esptool import main as esptool_main
+                import sys
+                from io import StringIO
+                
+                # Capture esptool output
+                old_stdout = sys.stdout
+                sys.stdout = captured_output = StringIO()
+                
+                try:
+                    # Run esptool image_info command
+                    sys.argv = ['esptool.py', 'image_info', '--version', '2', temp_file.name]
```

Comment:
Is there a reason we have to call it in this process like this? Why can't we do a subprocess call?